### PR TITLE
Feature/spring boot 2.x accumulo 2.1

### PR DESCRIPTION
**Pull request by jwomeara**

### CODE REVIEW

File: api/src/main/java/datawave/webservice/metadata/MetadataFieldBase.java [hunk 1]
```diff
@@ -11,6 +11,7 @@
 import java.util.List;
 import java.util.Set;
 
+// TODO if nothing other than DefaultMetadataField extends this class, remove this class and rename DefaultMetadataField to MetadataField
```
Comment (matthpeterson) [inline api/src/main/java/datawave/webservice/metadata/MetadataFieldBase.java]: Was this TODO left intentionally?

File: api/src/main/java/datawave/webservice/model/FieldMapping.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    public int compareTo(FieldMapping obj) {
+        
+        if (obj == null) {
+            throw new IllegalArgumentException("can not compare null");
+        }
+        
+        if (obj == this)
+            return 0;
+        
+        return new CompareToBuilder().append(datatype, ((FieldMapping) obj).datatype).append(fieldName, ((FieldMapping) obj).fieldName)
```
Comment (matthpeterson) [inline api/src/main/java/datawave/webservice/model/FieldMapping.java]: ignorable nit: casting obj once and placing into a variable would improve readability.

Comment (matthpeterson) [reply api/src/main/java/datawave/webservice/model/FieldMapping.java]: on second thought, the casting isn't necessary at all

File: service/src/main/java/datawave/microservice/metadata/DefaultMetadataFieldScanner.java [hunk 3]
```diff
@@ -272,7 +271,7 @@ private void setType() {
         
         // Set the last updated date for the current {@link DefaultMetadataField} based on the timestamp of the current entry.
         private void setLastUpdated() {
-            currField.setLastUpdated(DATE_FORMATTER.print(currKey.getTimestamp()));
+            currField.setLastUpdated(new SimpleDateFormat(TIMESTAMP_FORMAT).format(new Date(currKey.getTimestamp())));
```
Comment (matthpeterson) [inline service/src/main/java/datawave/microservice/metadata/DefaultMetadataFieldScanner.java]: Suggestion:
```
LocalDate.now().format(DateTimeFormatter.ofPattern(TIMESTAMP_FORMAT));
```
https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ofPattern-java.lang.String-
https://martin-grigorov.medium.com/compare-performance-of-javas-simpledateformat-against-datetimeformatter-31be58cadf1d

File: service/src/main/java/datawave/microservice/model/ModelController.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    
+    private final String dataTablesUri;
+    private final String jqueryUri;
+    private final AccumuloConnectionService accumloConnectionService;
+    
+    public static final String DEFAULT_MODEL_TABLE_NAME = "DatawaveMetadata";
+    private static final HashSet<String> RESERVED_COLF_VALUES = Sets.newHashSet("e", "i", "ri", "f", "tf", "m", "desc", "edge", "t", "n", "h");
+    
+    public ModelController(ModelProperties modelProperties, AccumuloConnectionService accumloConnectionService) {
+        this.dataTablesUri = modelProperties.getDefaultTableName();
```
Comment (matthpeterson) [inline service/src/main/java/datawave/microservice/model/ModelController.java]: Looks like this should be ```this.dataTablesUri = modelProperties.getDataTablesUri();```

File: service/src/main/java/datawave/microservice/model/ModelController.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+     * @param modelTableName
+     *            name of the table that contains the model
+     * @return a VoidResponse
+     * @RequestHeader X-ProxiedEntitiesChain use when proxying request for user
+     *            
+     * @HTTP 200 success
+     * @HTTP 500 internal server error
+     */
+    @Operation(summary = "Insert a new field mapping into an existing model.")
+    @PostMapping(value = {"/insert", "/import"}) // If we get to change to standard REST, this would just be / and rely on the
```
Comment (matthpeterson) [inline service/src/main/java/datawave/microservice/model/ModelController.java]: unfinished comment

File: service/src/main/java/datawave/microservice/dictionary/data/DataDictionary.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    
+    void setDescriptions(Connection connectionConfig, String fieldName, String datatype, Set<DESC> descriptions) throws Exception;
+    
+    Multimap<Entry<String,String>,DESC> getDescriptions(Connection connectionConfig) throws Exception;
+    
+    Multimap<Entry<String,String>,DESC> getDescriptions(Connection connectionConfig, String datatype) throws Exception;
+    
+    Set<DESC> getDescriptions(Connection connectionConfig, String fieldName, String datatype) throws Exception;
+    
+    void deleteDescription(Connection connectionConfig, String fieldName, String datatype, DESC description) throws Exception;
```
Comment (matthpeterson) [inline service/src/main/java/datawave/microservice/dictionary/data/DataDictionary.java:35]: it seems odd that description is supplied here.  it looks like it's only used to determine the col vis, which I guess is ok

File: service/src/main/java/datawave/microservice/AccumuloConnectionService.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+     * @param requestedAuthorizations
+     *            the level of authorizations to downgrade to
+     * @param currentUser
+     *            the user for whom to downgrade the authorizations
+     *            
+     * @return the authorizations for the specified user downgraded to the specified level
+     */
+    public Set<Authorizations> getDowngradedAuthorizations(String requestedAuthorizations, DatawaveUserDetails currentUser) {
+        DatawaveUser primaryUser = currentUser.getPrimaryUser();
+        return userAuthFunctions.mergeAuthorizations(userAuthFunctions.getRequestedAuthorizations(requestedAuthorizations, currentUser.getPrimaryUser()),
```
Comment (matthpeterson) [inline service/src/main/java/datawave/microservice/AccumuloConnectionService.java]: second call to currentUser.getPrimaryUser() could use primaryUser variable.

File: service/src/main/java/datawave/microservice/AccumuloConnectionService.java [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+     *            the regex (can be just a string) to apply to the IteratorSetting
+     *            
+     * @return the keys of hte specified table that the specified user has the authorzations to access.
+     * @throws TableNotFoundException
+     *             Thrown if the table is not found
+     */
+    public List<Key> getKeys(String modelTable, DatawaveUserDetails currentUser, String regexTerm) throws TableNotFoundException {
+        Scanner scanner = ScannerHelper.createScanner(this.accumuloClient, modelTable, this.getAuths(currentUser));
+        if (!regexTerm.isEmpty()) {
+            IteratorSetting cfg = new IteratorSetting(21, "colfRegex", RegExFilter.class.getName());
```
Comment (matthpeterson) [inline service/src/main/java/datawave/microservice/AccumuloConnectionService.java:169]: I wonder how many of the calls to this require a regex and if they can be replaced with an exact colf value which could be more performant, depending on locality groups.

File: api/src/main/java/datawave/webservice/metadata/MetadataFieldBase.java (hunk 1 above)
Comment (jwomeara) [reply api/src/main/java/datawave/webservice/metadata/MetadataFieldBase.java]: Good question.  This was a change Ceria added, so I'm not sure.  I'm going to remove the comment.

File: api/src/main/java/datawave/webservice/model/FieldMapping.java (hunk 2 above)
Comment (jwomeara) [reply api/src/main/java/datawave/webservice/model/FieldMapping.java]: I agree.  This isn't a new change.  This is another file copied over from here: https://github.com/NationalSecurityAgency/datawave/blob/integration/web-services/client/src/main/java/datawave/webservice/model/FieldMapping.java

I'm removing the unnecessary casts in the next commit.

File: service/src/main/java/datawave/microservice/AccumuloConnectionService.java (hunk 7 above)
Comment (jwomeara) [reply service/src/main/java/datawave/microservice/AccumuloConnectionService.java]: agreed.  fixing in next commit.

File: service/src/main/java/datawave/microservice/AccumuloConnectionService.java (hunk 8 above)
Comment (jwomeara) [reply service/src/main/java/datawave/microservice/AccumuloConnectionService.java:169]: I think this is representative of the original logic here: https://github.com/NationalSecurityAgency/datawave/blob/integration/web-services/model/src/main/java/datawave/webservice/query/model/ModelBean.java#L356

This was part of some refactor work Ceria did to move the model rest endpoint to the dictionary service.

File: service/src/main/java/datawave/microservice/dictionary/data/DataDictionary.java (hunk 6 above)
Comment (jwomeara) [reply service/src/main/java/datawave/microservice/dictionary/data/DataDictionary.java:35]: Sorry to keep making these same kinds of comments, but this isn't new, just a relocation of the following class during Ceria's refactor: https://github.com/NationalSecurityAgency/datawave-dictionary-service/blob/main/service/src/main/java/datawave/microservice/dictionary/data/DatawaveDataDictionary.java

File: service/src/main/java/datawave/microservice/metadata/DefaultMetadataFieldScanner.java (hunk 3 above)
Comment (jwomeara) [reply service/src/main/java/datawave/microservice/metadata/DefaultMetadataFieldScanner.java]: change incoming with the next commit.  let me know if that's better.

File: service/src/main/java/datawave/microservice/model/ModelController.java (hunk 4 above)
Comment (jwomeara) [reply service/src/main/java/datawave/microservice/model/ModelController.java]: good catch!

File: service/src/main/java/datawave/microservice/model/ModelController.java (hunk 5 above)
Comment (jwomeara) [reply service/src/main/java/datawave/microservice/model/ModelController.java]: im not sure what was meant by that comment.  deleted.

**Review by matthpeterson — DISMISSED** [review]

**Review by apmoriarty — DISMISSED** [review]

**Review by nonessentialprototype — DISMISSED** [review]

**Review by matthpeterson — APPROVED** [review]

**Pull request merged**